src/java.base/share/classes/sun/security/ssl/SSLSessionContextImpl.java

```diff
@@ -170,6 +178,52 @@ public int getSessionCacheSize() {
         return cacheLimit;
     }
 
+    // Package-private, used only from SSLContextImpl::engineInit() to initialie currentKeyID.
+    void initCurrentKeyID(int keyID) {
+        currentKeyID = keyID;
+    }
+
+    // Package-private, used only from SSLContextImpl::getKey() to create a new session key.
+    int getCurrentKeyID() {
+        return currentKeyID;
+    }
+
+    private void cleanupSessionKeys() {
+        Iterator<Map.Entry<Integer, SessionTicketExtension.StatelessKey>> it =
+            keyHashMap.entrySet().iterator();
+        while (it.hasNext()) {
+            Map.Entry<Integer, SessionTicketExtension.StatelessKey> entry = it.next();
+            SessionTicketExtension.StatelessKey k = entry.getValue();
+            if (k.isInvalid(this)) {
+                it.remove();
+                try {
+                    k.key.destroy();
+                } catch (Exception e) {
+                    // Suppress
+                }
+            }
+        }
+    }
+
+    // Every time we insert a new session key we check for and delete invalid keys.
+    // Package-private, used only from SSLContextImpl::getKey() to create a new session key.
+    void insertNewSessionKey(int newID, SessionTicketExtension.StatelessKey ssk) {
+        assert newID != currentKeyID : "Must use a new ID for a new session key!";
+        keyHashMap.put(Integer.valueOf(newID), ssk);
+        currentKeyID = newID;
+        cleanupSessionKeys();
+    }
+
+    // Package-private, used only from SSLContextImpl::getKey().
+    SessionTicketExtension.StatelessKey getKey() {
+        return keyHashMap.get(currentKeyID);
+    }
+
+    // Package-private, used only from SSLContextImpl::getKey(int id).
+    SessionTicketExtension.StatelessKey getKey(int id) {
+        return keyHashMap.get(id);
+    }
+
```

Comment:
It may need to reconsider these 3 methods you want to move SSLContextImpl.getKey() into this class.